### DISCUSSION TITLE
[Feature request] Allow changing to a locale which doesn't exist in initial config `supportedLocales`

### DISCUSSION THREAD

**Discussion by ahnpnl** [robisim74/qwik-speak#29] [General]
### Scenario

My application gets user config on load to get user's default locale and supported locales. I have a default `speakConfig` like below
```
{
     defaultLocale: { lang: 'en-US' },
    supportedLocales: [{ lang: 'en-US' }],
    assets: ['common'],
}
```

I tried 2 attempts:

- Creating a wrapper component around `QwikSpeakProvider` and pass a reactive store data into my wrapper component.  The reactive store data is set based on API response. However, this way doesn't work.
- Manually set value via `useSpeakContext`. For example `speakContext.config.supportedLocales = [{ lang: 'en-US' }, { lang: 'nl-NL' }]`. This way doesn't work either

### Request

A new feature to allow config to be dynamic configured or if new feature is not possible, maybe some documentation updates?

**Comment by robisim74** [comment]
Hi @ahnpnl,

I blocked with `Object.freeze(config)` editing speak config dynamically for two reasons:
- making it reactive and updatable consumes unnecessary resources, because Qwik creates and serializes subscriptions
- in my opinion a configuration must be available before the app is rendered and shouldn't be changed at runtime (even if it's just a translation library)

About you first attempt, if you create a component to update a state in the parent component, then Qwik (yet for the moment) doesn't make the state immediately available, but only within the child components (note the usage of the `Speak` component in this library).

I think you should use `useTask$` with `isServer` (ex `useServerMount$`), i.e.:

**root.tsx**
```jsx
export default component$(() => {
  const urlEnv = useServerData<string>('url');
  const origin = isServer && urlEnv ? (new URL(urlEnv)).origin : '';

  let config: SpeakConfig = {
    defaultLocale: { lang: '' }, 
    supportedLocales: [],
    assets: []
  };

  useTask$(async () => {
    if (isServer) {
      const response = await fetch(`${origin}/config.json`);  // just an example
      const result: SpeakConfig = await response.json();
      config.defaultLocale = result.defaultLocale;
      config.supportedLocales = result.supportedLocales;
      config.assets = result.assets
    }
  });

  return (
    /**
     * Init Qwik Speak (only available in child components)
     */
    <QwikSpeakProvider config={config} translationFn={translationFn}>
      ...
    </QwikSpeakProvider>
  );
});
```

Another way could be to get the config server side (adapter), for example in node, and pass it as an option to the root component (at least in production) and use a local config in dev mode.

**Comment by ahnpnl** [comment]
Thanks, I think using `useTask$` with `isServer` would work since it will block rendering until the async api call is resolved. My 1st attempt actually tried to optimize by not blocking rendering but did streaming.

Thank you for your response, I think I would have to take the downside about blocking rendering.

**Comment by robisim74** [comment]
I tried to allow config dynamically configurable, just remove this line: https://github.com/robisim74/qwik-speak/blob/434c2ebf9e49751edf8bdb58199518780a9928bb/src/library/qwik-speak-component.tsx#L80  but I get this error from Qwik:

`Error: State mutation inside render function. Move mutation to useWatch(), useBrowserVisibleTask() or useServerMount()`

The message is not updated (`useWatch() + useServerMount() = useTask$`), but in the end Qwik doesn't seem to allow the change once the context is initialized during rendering.

**Reply by ahnpnl** [reply]
Hi @robisim74 , I've prepared a repo to compare `qwik-speak` features with `vue-i18n`. You can check out the repo here https://github.com/ahnpnl/monorepo-multi-apps

## Usage

- Install deps with `yarn`
- At root `package.json`, there are 2 scripts: `serve-vue` and `serve qwik` to start a Vue application and a Qwik application

## Use case explanation

In each application:
- There are 2 buttons `English` and `Nederlands` to change locale and load new translations.
- There are 2 texts to print out simple translation and pluralization translation.
- There are 2 texts to print out current locale and supported locales. Current locale and supported locales are application state.
- Open `Infrastructure/Translations/TranslationProvider.ts` or `Infrastructure/Translations/TranslationProvider.tsx` to see implementation.

## Differences

### Change locale

For `vue-i18n`, whenever changing locale, devs have to call `setLocaleMessage` to manually set the new messages which is resolved from `fetch` while `qwik-speak` does it under the hood.

However, because of this under the hood logic, `qwik-speak` prevents devs from setting a new locale if that locale is not provided as `supportedLocales` in default config. This is the problem I encounter in my real app where I fetch user locales via API and I want to change to one of the user's locales later.

With `vue-i18n`, they don't have the concept of `supportedLocales` in initial config, therefore I think they allow to change to a new locale easier. 

### Pluralization

I checked the documentation and I couldn't find possibility to do pluralization. Is it a missing feature? You can check example from `vue-i18n` here https://vue-i18n.intlify.dev/guide/essentials/pluralization.html

### Fallback locale

One thing I like from `vue-i18n` which I think `qwik-speak` can adopt is the possibility to configure custom `fallbackLocale`.

### Initial messages

In `vue-i18n`, it's possible to define initial messages which allow devs to import `json` files and put them into initial config. Idk if this feature `qwik-speak` can adopt too, maybe something to check?

I hope this gives you some more insights about my use case. Thank you.

**Reply by robisim74** [reply]
@ahnpnl Thanks for your comparison repo. But when making this comparison, we need to keep these things in mind:
- Qwik (and this library) are still in beta, while Vue and vue-i18n have been around for many years and are stable
- Qwik works very differently from Vue
- This library is completely different from vue-i18n: maybe it looks like libraries based on i18n-next and Intl
- This library started as a side project as my need, I decided to share it because there was no i18n support in Qwik yet
- The main goal of this library is to provide a structure for localization with smallest possible dimensions (at the moment about 8 kB, what is the size of the vue-18n?), because the real localization has to happen in the build phase, according to conclusions with Miško Hevery in this thread: https://github.com/robisim74/qwik-speak/discussions/8 and with Qwik i18n docs: https://qwik.builder.io/docs/advanced/i18n/. Only Angular `$localize` (framework independent and suggested by Miško  himself) applies the same logic

You are raising many questions, which I will try to answer, but which will certainly need to be investigated:
- **Change locale**: When you call `changeLocale `function, it set in the state the new locale. The locale is reactive, and the library call your custom `loadTranslation$` implementation to load the translations for the new locale at runtime. As you can find in Qwik documentation https://qwik.builder.io/docs/advanced/i18n/, loading files at runtime is only recommended in dev mode, while in production it is recommended to inline translations
- **supportedLocales** are indispensable in this library because `SpeakLocale `is different from the simple `languageCode-countryCode` (see [SpeakLocale](https://github.com/robisim74/qwik-speak#speaklocale)). But there are situations where only `languageCode-countryCode` is available (for example the lang parameter in the url, and it would not be possible to bind for a default value for currency or timezone for example). Normally the locales supported by the app are known, even if loaded via API. By _locales supported by the app_ I mean the locales for which the app has the translation files available
- **Fallback locale** `defaultLocale` in this library looks like it (maybe we need to rename it?): if a locale is not supported, the `defaultLocale` is used. The difference with vue-i18n and with all i18n libraries is that it is not a fallback on the single value that is not found, but on the whole locale. Why? Again for the reason of localization during the build. When inlining, _the translations must already have merged the fallback value_, using the `defaultValue `in the` $translate` function and doing the extraction: see [extract](https://github.com/robisim74/qwik-speak/blob/main/tools/extract.md) and [inline](https://github.com/robisim74/qwik-speak/blob/main/tools/inline.md) 
- **Initial messages**: I'm not sure I understand: `loadTranslation$` is already called by `QwikSpeakProvider `with the `assets `provided in configuration for the resolved locale when the app is initialized
- **Pluralization**: At the moment there is the `$plural` function. It is based on Intl API, like number and date functions. It's not well documented yet, you are right

**About your original question** (dynamically loading config or supporting locales not in `supportedLocales`). Initially this library offered other custom functions besides` loadTranslation$`, such as `resolveLocale$`: unfortunately I had to remove them, because Qrl functions don't offer many possibilities (like access to State or Request), and I've given the option to resolve locale and configuration before passing them to the `QwikSpeakProvider `(which needs them right away, and they can't be changed, because the rendering takes place on the server and is sent to the browser). Maybe I'm missing something, but I don't have any better solutions right now.

**Reply by ahnpnl** [reply]
Thank you for your response.

I agree that app supported locales is often a bigger set than user’s locales so that would be the way to go given the difference between vue & qwik. I think the current implementation is already good for that.

I think the only thing left is documentation about how to translate with pluralization.

**Reply by robisim74** [reply]
Added: https://robisim74.gitbook.io/qwik-speak/library/translate#usdplural

**Reply by ahnpnl** [reply]
Thanks 🙏

**Reply by ahnpnl** [reply]
Hi @robisim74

I just noticed that `$plural` API doesn't support param interpolation. Is it something missing or I misunderstand the API? I have the json
```
"devs": {
    "one": "{{ value }} software developer {{ name }}",
    "other": "{{ value }} software developers {{ name }}"
  }
```
and when I look at `$plural` API, there is no way to pass value for `name`.

Another small thing is, I think `prefix` param name in `$plural` should be renamed to `key` to be clearer.
